source/common/io/io_uring_worker.h

```diff
@@ -0,0 +1,509 @@
+      pending_result_ = pending_read_buf_.length();
+    }
+
+    // ASSERT(pending_read_buf_.length() == 0);
+    if (read_req_ == nullptr && !is_closing_ && !is_disabled_) {
+      ENVOY_LOG(trace, "submit read request after previous read complete, fd = {}", fd_);
+      submitReadRequest();
+    }
+  }
+
+  void onWrite(int32_t result, Request* req) override {
+    ENVOY_LOG(trace, "onWrite with result {}, fd = {}", result, fd_);
+    bool injected_event = (result == -EAGAIN && write_req_ != req);
+    if (injected_event) {
+      if (is_closing_) {
+        ENVOY_LOG(trace, "there is a inject event, but is closing, ignore it, fd = {}", fd_);
+        return;
+      }
+      // TODO: using ptr in read param;
+      ENVOY_LOG(trace,
+                "there is a inject event, and same time we have regular write request, fd = {}",
+                fd_);
+      WriteParam param{result};
+      io_uring_handler_.onWrite(param);
+      return;
+    }
+
+    write_req_ = nullptr;
+    if (iovecs_ != nullptr) {
+      delete[] iovecs_;
+      iovecs_ = nullptr;
+    }
+
+    if (result <= 0) {
+      if (is_closing_ && read_req_ == nullptr && cancel_req_ == nullptr && write_req_ == nullptr &&
+          close_req_ == nullptr) {
+        ASSERT(close_req_ == nullptr);
+        close_req_ = parent_.submitCloseRequest(*this);
+        return;
+      }
+
+      if (result == -EAGAIN) {
+        if (is_closing_) {
+          ENVOY_LOG(trace, "ignore injected event when closing");
+          return;
+        }
+        if (!injected_event) {
+          if (write_buf_.length() > 0) {
+            ENVOY_LOG(trace, "continue write buf since get eagain, size = {}, fd = {}",
+                      write_buf_.length(), fd_);
+            submitWriteRequest();
+          }
+          return;
+        }
+      }
+
+      WriteParam param{result};
+      io_uring_handler_.onWrite(param);
+      return;
+    }
+
+    write_buf_.drain(result);
+    ENVOY_LOG(trace, "drain write buf, drain size = {}, fd = {}", result, fd_);
+    if (write_buf_.length() > 0) {
+      ENVOY_LOG(trace, "continue write buf since still have data, size = {}, fd = {}",
+                write_buf_.length(), fd_);
+      submitWriteRequest();
+      return;
+    }
+
+    if (is_closing_ && read_req_ == nullptr && cancel_req_ == nullptr) {
+      ENVOY_LOG(trace, "write is done, try to close the socket, fd = {}", fd_);
+      ASSERT(close_req_ == nullptr);
+      close_req_ = parent_.submitCloseRequest(*this);
+      return;
+    }
+
+    if (is_full_ && write_buf_.length() == 0) {
+      is_full_ = false;
+      WriteParam param{result};
+      io_uring_handler_.onWrite(param);
+    }
+  }
+
+  uint64_t write(Buffer::Instance& buffer) override {
+    ENVOY_LOG(trace, "write, size = {}, fd = {}, slices = {}", buffer.length(), fd_,
+              buffer.getRawSlices().size());
+    uint64_t buffer_limit = 1024 * 1024;
+    uint64_t writable_size = buffer_limit - write_buf_.length();
```

Comment:
what prevents this from underflowing?

Reply:
emm...good question. There should be no limit. Initially, I just thought the buffer should have a limit. But the envoy up layer code should already limit the buffer size. There won't be a case for unlimited huge size to write.